### PR TITLE
2.6: Development process and tools

### LINKED ISSUE #214: 2.6 - Process Development Process & Tools

### PR DESCRIPTION
**Pull request by jlndk**

### CODE REVIEW

File: report/26_Process-Development-Process-&-Tools.md [hunk 1]
```diff
@@ -1,2 +1,52 @@
...
+
+<!-- 
+**TODO Discord**
+
+* Channels
+* Voice chats
+* Bots 
+-->
+
+**TODO: Descripe how we where devops when writing report**
```
Comment (jlndk) [inline report/26_Process-Development-Process-&-Tools.md]: Dont know if this actually fits here or somewhere else

Comment (Hjaltesorgenfrei) [reply report/26_Process-Development-Process-&-Tools.md]: It fits here.

File: report/26_Process-Development-Process-&-Tools.md [hunk 2]
```diff
@@ -1,2 +1,52 @@
-Applied development process and tools supporting it
-For example, how did you use issues, Kanban boards, etc. to organize open tasks
+### Applied development process and tools supporting it
+
+Throughout this course, two tools, Github and Discord, has been used to manage the project and communication.
+Two tools have been used, to differentiate between synchronous and asynchronous communication.
```
Comment (Hjaltesorgenfrei) [inline report/26_Process-Development-Process-&-Tools.md]: > Two tools have been used, to differentiate between synchronous and asynchronous communication.
Thats not correct.
synchronous communication is talking. Discord is async

File: report/26_Process-Development-Process-&-Tools.md [hunk 3]
```diff
@@ -1,2 +1,52 @@
...
+That is, Discord is used to communicate directly and informally to allow for fast and efficient communication,
+whereas GitHub is used to organize our work in a way that can easily be searched through and understood at a later date.
+
+GitHub has been the ideal tool for the majority of our project management, because both the code, git commits, and corresponding issues coexist and can easily be cross-referenced.
+Furthermore, GitHub has the concept of "Pull requests" (PR), which introduces a way to do formal review policy for new code, before merging it with the master branch.
+These pull requests can also be linked with the aforementioned issues, which makes it easy to see which code-changes fixes a given issue.
```
Comment (Hjaltesorgenfrei) [inline report/26_Process-Development-Process-&-Tools.md]: Add a description of why this is a good thing.

File: report/26_Process-Development-Process-&-Tools.md [hunk 4]
```diff
@@ -1,2 +1,52 @@
...
+
+The Github workflow starts by creating the issues. The actual content of the issues is often agreed upon verbally or dictated by the homework for the course. The issues are then augmented with relevant metadata, such as description, tags, milestones, assignees, and comments. Most importantly, we use a tag for all mandatory work, so that we know what needs to be completed.
```
Comment (Hjaltesorgenfrei) [inline report/26_Process-Development-Process-&-Tools.md]: issues -> issue

Comment (Hjaltesorgenfrei) [reply report/26_Process-Development-Process-&-Tools.md]: homework -> assignment from the course

File: report/26_Process-Development-Process-&-Tools.md [hunk 5]
```diff
@@ -1,2 +1,52 @@
...
+Occasionally, when many tasks need to be solved in a certain timeframe, such as when writing this report, we make a "milestone" in GitHub.
```
Comment (Hjaltesorgenfrei) [inline report/26_Process-Development-Process-&-Tools.md]: remove all the "we"s

File: report/26_Process-Development-Process-&-Tools.md [hunk 6]
```diff
@@ -1,2 +1,52 @@
...
+<!-- 
+**TODO GitHub**:
+* Issues
+* PR's
+* Reviews
+* Labels
+* Milestones 
+-->
+
+Discord has been used where GitHub falls short, namely for instant synchronous communication.
```
Comment (Jakobis) [inline report/26_Process-Development-Process-&-Tools.md]: "instant synchronous" seems like a dobbeltkonfekt

File: report/26_Process-Development-Process-&-Tools.md [hunk 7]
```diff
@@ -1,2 +1,52 @@
...
+
+GitHub has been the ideal tool for the majority of our project management, because both the code, git commits, and corresponding issues coexist and can easily be cross-referenced.
+Furthermore, GitHub has the concept of "Pull requests" (PR), which introduces a way to do formal review policy for new code, before merging it with the master branch.
+These pull requests can also be linked with the aforementioned issues, which makes it easy to see which code-changes fixes a given issue.
+
+The Github workflow starts by creating the issues. The actual content of the issues is often agreed upon verbally or dictated by the homework for the course. The issues are then augmented with relevant metadata, such as description, tags, milestones, assignees, and comments. Most importantly, we use a tag for all mandatory work, so that we know what needs to be completed.
+Occasionally, when many tasks need to be solved in a certain timeframe, such as when writing this report, we make a "milestone" in GitHub.
+In this, we can attach all relevant issues, set a due date and afterward track the combined progress for all tasks.
```
Comment (Hjaltesorgenfrei) [inline report/26_Process-Development-Process-&-Tools.md]: Reword this entire sentence. Remove we, start with explaining that its a postive of milestones.

File: report/26_Process-Development-Process-&-Tools.md [hunk 8]
```diff
@@ -1,2 +1,52 @@
...
+When a developer is ready to solve the issue, they first assign themself if they were not assigned when the issue was created. Afterward, they make a branch and make the changes to the code, as described in (TODO: Refer to GitHub flow chapter). When the code change is complete, they create a new PR from their branch to master, and augment it with the relevant data, such as a descriptive title, description, and linked issues. By creating this process, it signals to the other developers that the code is ready for review. Furthermore, the developer also posts a link to the PR on discord to notify the rest of the team.
```
Comment (Hjaltesorgenfrei) [inline report/26_Process-Development-Process-&-Tools.md]: > When a developer is ready to solve the issue, they first assign themself if they were not assigned when the issue was created

akward to read.

Comment (Hjaltesorgenfrei) [reply report/26_Process-Development-Process-&-Tools.md]: Fix todo.

File: report/26_Process-Development-Process-&-Tools.md [hunk 9]
```diff
@@ -1,2 +1,52 @@
...
+* Issues
+* PR's
+* Reviews
+* Labels
+* Milestones 
+-->
+
+Discord has been used where GitHub falls short, namely for instant synchronous communication.
+Despite the "gamer-branding" we used Discord instead of other popular team communication tools, such as messenger or slack, for various reasons.
+Besides already using it extensively in our daily lives, Discord has a powerful feature set that makes it incredibly well suited for distributed teamwork.
```
Comment (Jakobis) [inline report/26_Process-Development-Process-&-Tools.md]: Feature set is a weird expressions, just write that it has powerful features

File: report/26_Process-Development-Process-&-Tools.md [hunk 10]
```diff
@@ -1,2 +1,52 @@
...
+Occasionally, when many tasks need to be solved in a certain timeframe, such as when writing this report, we make a "milestone" in GitHub.
+In this, we can attach all relevant issues, set a due date and afterward track the combined progress for all tasks.
+When a developer is ready to solve the issue, they first assign themself if they were not assigned when the issue was created. Afterward, they make a branch and make the changes to the code, as described in (TODO: Refer to GitHub flow chapter). When the code change is complete, they create a new PR from their branch to master, and augment it with the relevant data, such as a descriptive title, description, and linked issues. By creating this process, it signals to the other developers that the code is ready for review. Furthermore, the developer also posts a link to the PR on discord to notify the rest of the team.
+
+For a pull request to be merged it must comply with our policies.
```
Comment (Hjaltesorgenfrei) [inline report/26_Process-Development-Process-&-Tools.md]: Change this.

Comment (Hjaltesorgenfrei) [reply report/26_Process-Development-Process-&-Tools.md]: What even is our policies? We dont have something formal like that.

File: report/26_Process-Development-Process-&-Tools.md [hunk 11]
```diff
@@ -1,2 +1,52 @@
...
+If a pull request is rejected, the author can make changes and re-request reviews.
+When a pull request is merged, it will automatically be deployed to our production environment (TODO: Refer to CI/CD chapter), and mark all linked issues as solved.
+<!-- 
+**TODO GitHub**:
+* Issues
+* PR's
+* Reviews
+* Labels
+* Milestones 
+-->
```
Comment (Hjaltesorgenfrei) [inline report/26_Process-Development-Process-&-Tools.md]: No Todos.

File: report/26_Process-Development-Process-&-Tools.md (hunk 1 above)
Comment (Jakobis) [reply report/26_Process-Development-Process-&-Tools.md]: I have no idea what that sentence means

File: report/26_Process-Development-Process-&-Tools.md [hunk 12]
```diff
@@ -1,2 +1,52 @@
...
+When a developer is ready to solve the issue, they first assign themself if they were not assigned when the issue was created. Afterward, they make a branch and make the changes to the code, as described in (TODO: Refer to GitHub flow chapter). When the code change is complete, they create a new PR from their branch to master, and augment it with the relevant data, such as a descriptive title, description, and linked issues. By creating this process, it signals to the other developers that the code is ready for review. Furthermore, the developer also posts a link to the PR on discord to notify the rest of the team.
+
+For a pull request to be merged it must comply with our policies.
+First of all, all automatic tests and checks must complete successfully. This includes unit tests, end-to-end tests, static analysis, and more.
+Secondly, the PR must be approved by at least one other developer. Github supports code reviews as a part of the pull requests, so that other team members can, approve, reject, and suggest changes to a PR.
+The automatic tests help catch bugs and regressions the human reviewer would otherwise miss, and visa versa.
```
Comment (Hjaltesorgenfrei) [inline report/26_Process-Development-Process-&-Tools.md]: visa versa? Thats spoken word and also does not make sense here.

File: report/26_Process-Development-Process-&-Tools.md [hunk 13]
```diff
@@ -1,2 +1,52 @@
...
+Even though we cannot eliminate bugs completely, but the combination of automatic tools, combined with human common sense, is the best way of attempting to prevent them.
```
Comment (Hjaltesorgenfrei) [inline report/26_Process-Development-Process-&-Tools.md]: no we.

File: report/26_Process-Development-Process-&-Tools.md (hunk 2 above)
Comment (Jakobis) [inline report/26_Process-Development-Process-&-Tools.md]: You also write that two tools have been used right above. These two sentences right after each other sound super weird

File: report/26_Process-Development-Process-&-Tools.md (hunk 6 above)
Comment (Hjaltesorgenfrei) [inline report/26_Process-Development-Process-&-Tools.md]: It was mostly used for async, but yes it was also used for sync.

File: report/26_Process-Development-Process-&-Tools.md [hunk 14]
```diff
@@ -1,2 +1,52 @@
...
+* Labels
+* Milestones 
+-->
+
+Discord has been used where GitHub falls short, namely for instant synchronous communication.
+Despite the "gamer-branding" we used Discord instead of other popular team communication tools, such as messenger or slack, for various reasons.
+Besides already using it extensively in our daily lives, Discord has a powerful feature set that makes it incredibly well suited for distributed teamwork.
+One of the biggest advantages is that a discord server supports multiple text and voice channels. 
+This makes it easy to categorize and organize the chat. The voice chats also allow us to enter and leave at will, making it ideal for long-term group work, where people can join and leave as they please.
+Another big advantage is that discord has an easy-to-use API, with support for webhooks. This has been utilized by hooking up GitHub and our uptime-monitoring tool, OhDear, to separate channels, so that we get automatically get notified when new updates occur. This has allowed us to act very fast on code changes and downtime.
```
Comment (Hjaltesorgenfrei) [inline report/26_Process-Development-Process-&-Tools.md]: I think its worth mention we did not write these tools.

Comment (Hjaltesorgenfrei) [reply report/26_Process-Development-Process-&-Tools.md]: Also "code changes" has not meaning.

**Review by Hjaltesorgenfrei — CHANGES_REQUESTED** [review]
Please remove all parts which are not part of 2.6. It is unnescary clutter to the commit.

**Review by Jakobis — DISMISSED** [review]
Two things: 
First, this has a lot of over lap with developer interactions, and we need to figure out what information should go where. 
Second, we and our all over the place, I am still of the opinion that we should avoid them.

**Comment by Hjaltesorgenfrei** [comment]
> First, this has a lot of over lap with developer interactions, and we need to figure out what information should go where.

All overlap from here should be deleted. Developer interaction should be in developer interaction.

File: report/26_Process-Development-Process-&-Tools.md (hunk 6 above)
Comment (Jakobis) [reply report/26_Process-Development-Process-&-Tools.md]: Apparently called a pleonasm in english

File: report/26_Process-Development-Process-&-Tools.md [hunk 15]
```diff
@@ -1,2 +1,32 @@
...
+Occasionally, when many tasks need to be solved in a certain timeframe, such as when writing this report, a "milestone" is created in GitHub, where it is possible to attach all relevant issues, set a due date, and afterward track the combined progress for all tasks.
+Before working on a task, the developer that is supposed to solve it, starts by assigning themself to the issue if they were not already assigned when the issue was created.
+Afterward, they make a branch and make the changes to the code, as described in (TODO: Refer to GitHub flow chapter). When the code change is complete, they create a new PR from their branch to master, and augment it with the relevant data, such as a descriptive title, description, and linked issues. By creating this process, it signals to the other developers that the code is ready for review. Furthermore, the developer also posts a link to the PR on discord to notify the rest of the team.
```
Comment (jlndk) [inline report/26_Process-Development-Process-&-Tools.md]: fix todo

File: report/26_Process-Development-Process-&-Tools.md [hunk 16]
```diff
@@ -1,2 +1,32 @@
...
+
+For a pull request to be merged it must comply with our policies.
+First of all, all automatic tests and checks must complete successfully. This includes unit tests, end-to-end tests, static analysis, and more.
+Secondly, the PR must be approved by at least one other developer. Github supports code reviews as a part of the pull requests, so that other team members can, approve, reject, and suggest changes to a PR.
+The automatic tests help catch bugs and regressions the human reviewer would otherwise miss, and the human usually catches things the automatic tests are not configured or able to detect.
+Even though it is essentially impossible to eliminate bugs completely, the combination of automatic tools, combined with human common sense, is the best way of attempting to prevent them.
+If a pull request is rejected, the author can make changes and re-request reviews.
+When a pull request is merged, it will automatically be deployed to our production environment (TODO: Refer to CI/CD chapter), and mark all linked issues as solved.
```
Comment (jlndk) [inline report/26_Process-Development-Process-&-Tools.md]: fix todo

File: report/26_Process-Development-Process-&-Tools.md [hunk 17]
```diff
@@ -1,2 +1,32 @@
...
+
+Discord has been used where GitHub falls short, namely for instant communication.
+Despite the "gamer-branding", Discord was used instead of other popular team communication tools, such as messenger or slack, for various reasons.
+Besides already using it extensively in our daily lives, Discord has many powerful features that make it incredibly well suited for distributed teamwork.
+One of the biggest advantages is that a discord server supports multiple text and voice channels.
+This makes it easy to categorize and organize the chat. The voice chats also allow us to enter and leave at will, making it ideal for long-term group work, where people can join and leave as they please.
+Another big advantage is that discord has an easy-to-use API, with support for webhooks. This has been utilized by hooking up GitHub and our uptime-monitoring tool, OhDear, to separate channels, so that notifications are automatically posted when new updates occur. This has allowed us to act very fast on code changes and downtime.
+
+**TODO: Describe how we used devops practices when writing this report**
```
Comment (jlndk) [inline report/26_Process-Development-Process-&-Tools.md]: fix todo

File: report/26_Process-Development-Process-&-Tools.md (hunk 14 above)
Comment (jlndk) [reply report/26_Process-Development-Process-&-Tools.md]: I might have fixed the "I think its worth mention we did not write these tools." parts by adding "existing" to the first sentence
> Throughout this course, two **existing** tools, Github and Discord, has been used to manage the project and communication.

**Review by Hjaltesorgenfrei — APPROVED** [review]

**Pull request merged**